### PR TITLE
Allow overriding haskell versions

### LINKED ISSUE #384: Allow version override for haskell dependencies
I imagine something like:

```typescript
mkHaskellProject({
  description: "something".
  ...
  versionOverrides: [
   { name: "servant-server",
     version: "0.17.1"
   },
   { name: "shake",
     github: "nmitchell/shake",
     commit: "k2jh3480012"
   }
  ]
})
```

Or 
```typescript
mkHaskellProject({
  description: "something".
  ...
  versionOverrides: {
   "servant-server": "0.17.1",
   "shake": {
      github: "nmichell/shake",
      commit: "..."
   }
})

### PR DESCRIPTION
**Pull request by jkarni**
Supersedes #420. Fixes #384

### CODE REVIEW

File: ts/haskell/mod.test.ts [hunk 1]
```diff
@@ -1,8 +1,60 @@
...
+        src: ".",
+        executables: ["foo"],
+        overrideDependencies: {
+          "string-conversions": "0.4.0.1",
+        },
+      });
+      output = assertSuccess(runExecutable(project.foo, { cwd: path }));
+      assertStdout(output, ">= 0.4\n");
+    });
+  });
```
Comment (soenkehahn) [inline ts/haskell/mod.test.ts:56]: I pushed fbe02289 as a review commit. Personally I find it easier to have the two project examples clearly separated. And re-using variables is also ok, although I care less about that.

File: ts/haskell/mod.ts [hunk 2]
```diff
@@ -37,23 +37,40 @@ type Execs<Exe extends string[]> = {
...
   description: string;
   compiler: string;
   executables?: Executables;
+  overrideDependencies?: Record<string, string>;
   src: string;
 }): Project & HaskellAddenda & Execs<Executables> {
-  const pkg: Package = mkPackage(
-    nixRaw`
-      (pkgs.haskell.packages.${nixRaw(args.compiler)}.callCabal2nix
+  const deps = args.overrideDependencies ? args.overrideDependencies : {};
```
Comment (soenkehahn) [inline ts/haskell/mod.ts:50]:
```suggestion
  const deps = args.overrideDependencies ?? {};
```

File: ts/haskell/mod.ts [hunk 3]
```diff
@@ -37,23 +37,38 @@ type Execs<Exe extends string[]> = {
...
+
+  const nixPkg = nixRaw`
+    let
+      haskell = pkgs.haskell.packages.${nixRaw(args.compiler)}.override {
+        overrides = final: prev:
+          ${nixAttrSet(
+            mapValues(
```
Comment (soenkehahn) [inline ts/haskell/mod.ts:57]: Pushed 2dfd9f7c, which I think is better. TIOLI.

**Review by soenkehahn — APPROVED** [review]
Very nice! Didn't know about `callHackage`. Does that just have a database of everything on hackage?

File: ts/haskell/mod.ts (hunk 3 above)
Comment (jkarni) [reply ts/haskell/mod.ts:57]: Much better!

File: ts/haskell/mod.test.ts [hunk 4]
```diff
@@ -1,38 +1,96 @@
...
+        },
+      });
+      const output1 = runExecutable(project1.foo, { cwd: path });
+      const output2 = runExecutable(project2.foo, { cwd: path });
+      assertSuccess(output1);
+      assertStdout(output1, "< 0.4\n");
+      assertSuccess(output2);
+      assertStdout(output2, ">= 0.4\n");
+    });
+  }),
```
[the remark this answers is not in the document]
Comment (jkarni) [reply ts/haskell/mod.test.ts]: This seems to have been fixed?

**Comment by jkarni** [comment]
> Very nice! Didn't know about callHackage. Does that just have a database of everything on hackage?

Yup. On that note, I feel like we should be updating nixpkgs on every release? Though currently things (like node and go version) are hard coded and might break.

**Pull request merged**